common/switchboard.hpp

```diff
@@ -1,179 +1,571 @@
-		return std::move(std::unique_ptr<reader_latest<event>>(reinterpret_cast<reader_latest<event>*>(void_writer.release())));
-	}
-
-	virtual ~switchboard() { }
-
-	virtual void stop() = 0;
-};
+    /**
+     * @brief A handle which can read the latest event on a topic.
+     */
+    template <typename specific_event>
+    class reader {
+    private:
+        /// Reference to the underlying topic
+        topic& _m_topic;
 
-/* TODO: (usability) Do these HAVE to be smart pointers? If the
-   copy-constructor is already shallow, they could be concrete
-   data-types. */
+    public:
+        reader(topic& topic_)
+            : _m_topic{topic_}
+        {
+#ifndef NDEBUG
+            if (typeid(specific_event) != _m_topic.ty()) {
+                std::cerr << "topic '" << _m_topic.name() << "' holds type " << _m_topic.ty().name()
+                          << ", but caller used type" << typeid(specific_event).name() << std::endl;
+                abort();
+            }
+#endif
+        }
 
-}
+       /**
+        * @brief Gets a "read-only" copy of the latest value.
+        *
+        * This will return null if no event is on the topic yet.
+        */
+       ptr<const specific_event> get_ro_nullable() const noexcept {
+          ptr<const event> this_event = _m_topic.get();
+          ptr<const specific_event> this_specific_event = std::dynamic_pointer_cast<const specific_event>(this_event);
+
+           if (this_event != nullptr) {
+			   assert(this_specific_event /* Otherwise, dynamic cast failed; dynamic type information could be wrong*/);
+               return this_specific_event;
+           } else {
+               return ptr<const specific_event>{nullptr};
+           }
+       }
+
+       /**
+        * @brief Gets a non-null "read-only" copy of the latest value.
+        *
+        * @throws `runtime_error` If no event is on the topic yet.
+        */
+        ptr<const specific_event> get_ro() const {
+           ptr<const specific_event> this_specific_event = get_ro_nullable();
+           if (this_specific_event != nullptr) {
+               return this_specific_event;
+           } else {
+               /// Otherwise, no event on the topic yet
+               throw std::runtime_error("No event on topic");
+           }
+        }
+
+       /**
+        * @brief Gets a non-null mutable copy of the latest value.
+        *
+        * @throws `runtime_error` If no event is on the topic yet.
+        */
+        ptr<specific_event> get_rw() const {
+           /*
+             This method is currently not more efficient than calling get_ro() and making a copy,
+             but in the future it could be.
+            */
+           ptr<const specific_event> this_specific_event = get();
+           return std::make_shared<specific_event>(*this_specific_event);
+        }
+
+        /// Member function alias for common case `get_ro`
+        const std::function< ptr<const specific_event>() > get = [this]() { return get_ro(); };
```

Comment:
My original intention was that `get()` was the default. Everyone should use that unless they had a reason to otherwise.. But then we decided to be more explicit.

Reply:
Removed.